src/main/kotlin/org/rust/lang/utils/RsDiagnostic.kt

```diff
@@ -156,6 +163,48 @@ sealed class RsDiagnostic(
         private fun expectedFound(expectedTy: Ty, actualTy: Ty): String {
             return "expected `${expectedTy.escaped}`, found `${actualTy.escaped}`"
         }
+
+        /**
+         * Try to find a "path" from [actualTy] to [expectedTy] through dereferences and references.
+         *
+         * The method works by getting coercion sequence of types for the [actualTy] and sequence of the types that can
+         * lead to [expectedTy] by adding references. The "expected sequence" is represented by a list of references'
+         * mutabilities, and a map from the type X to the index i in the list such that if we apply the references from
+         * 0 (inclusive) to i (exclusive) we will get [expectedTy]. For example for [expectedTy] = `&mut&i32` we will
+         * have: [mutable, immutable] and {&mut&i32 -> 0, &i32 -> 1, i32 -> 2}. Then, using those data structures, we
+         * try to find a first type X of the "actual" sequence that is also in the "expected" sequence, and a number of
+         * dereferences to get from [actualTy] to X and references to get from X to [expectedTy]. Finally we try to
+         * check if applying the references would agree with the expression's mutability.
+         */
+        private fun derefRefPathFromActualToExpected(lookup: ImplLookup, element: RsElement): DerefRefPath? {
+            // get all the types that can lead to `expectedTy` by adding references to them
+            val expectedRefSeq: MutableList<Mutability> = mutableListOf()
+            val tyToExpectedRefSeq: Map<Ty, Int> = buildMap {
+                put(expectedTy, 0)
+                var ty = expectedTy
+                var i = 1
+                while (ty is TyReference) {
+                    expectedRefSeq.add(ty.mutability)
+                    ty = ty.referenced
+                    put(ty, i++)
+                }
+            }
+            // get all the types we can get by dereferencing `actualTy`
+            val actualCoercionSeq = lookup.coercionSequence(actualTy).toList()
+            var refSeqEnd: Int? = null
+            // for the first type X in the "actual sequence" that is also in the "expected sequence"; get the number of
+            // dereferences we need to apply to get to X from `actualTy` and number of references to get to `expectedTy`
+            val derefs = actualCoercionSeq.indexOfFirst { refSeqEnd = tyToExpectedRefSeq.get(it); refSeqEnd != null }
+            val refs = expectedRefSeq.subList(0, refSeqEnd?: return null)
+            // check that mutability of references would not contradict the `element`
+            if (!refs.isEmpty() && refs.last() == Mutability.MUTABLE && (!isMut(element) ||
```

Comment:
Hmm, this condition definitely should be simplified/reformatted somehow. I think, we can extract it to a variable and invert it:
```kotlin
val isSuitableMutability = refs.isEmpty() || !refs.last().isMut || element.isMutable &&
        // covers cases lke `let mut x: &T = ...`
        actualCoercionSeq.subList(0, derefs + 1).all {
            it !is TyReference || it.mutability.isMut
        }
if (!isSuitableMutability) return null
```

Reply:
cool, done